haystack/pipeline_eval.py

```diff
@@ -0,0 +1,190 @@
+    #  maybe per sample results should be passed on to a final node which does all the aggregation
+    #  and metric computation
+    print_eval_metrics(eval_retriever, eval_reader)
+
+def print_eval_metrics(eval_retriever, eval_reader):
+    total_queries = eval_retriever.query_count
+    retriever_recall = eval_retriever.recall
+    correct_retrieval = eval_retriever.correct_retrieval
+    reader_top_1_em = eval_reader.top_1_em
+    reader_top_k_em = eval_reader.top_k_em
+    reader_top_1_f1 = eval_reader.top_1_f1
+    reader_top_k_f1 = eval_reader.top_k_f1
+    pipeline_top_1_em = eval_reader.top_1_em_count / total_queries
+    pipeline_top_k_em = eval_reader.top_k_em_count / total_queries
+    pipeline_top_1_f1 = eval_reader.top_1_f1_sum / total_queries
+    pipeline_top_k_f1 = eval_reader.top_k_f1_sum / total_queries
+
+    print("Retriever")
+    print("-----------------")
+    print(f"total queries: {total_queries}")
+    print(f"recall: {retriever_recall}")
+    print()
+    print("Reader")
+    print("-----------------")
+    print(f"answer in retrieved docs: {correct_retrieval}")
+    print(f"top 1 EM: {reader_top_1_em}")
+    print(f"top k EM: {reader_top_k_em}")
+    print(f"top 1 F1: {reader_top_1_f1}")
+    print(f"top k F1: {reader_top_k_f1}")
+    print()
+    print("Pipeline")
+    print("-----------------")
+    print(f"top 1 EM: {pipeline_top_1_em}")
+    print(f"top k EM: {pipeline_top_k_em}")
+    print(f"top 1 F1: {pipeline_top_1_f1}")
+    print(f"top k F1: {pipeline_top_k_f1}")
+
+class EvalRetriever:
+    def __init__(self):
+        self.outgoing_edges = 1
+        self.correct_retrieval = 0
+        self.query_count = 0
+        self.recall = 0.0
+        # self.log = []
+
+    def run(self, documents, labels, **kwargs):
+        # Open domain mode
+        self.query_count += 1
+        if type(labels) == str:
+            labels = [labels]
+        texts = [x.text for x in documents]
+        correct_retrieval = False
+        for t in texts:
+            for label in labels:
+                if label.lower() in t.lower():
+                    self.correct_retrieval += 1
+                    correct_retrieval = True
+                    break
+            if correct_retrieval:
+                break
+        self.recall = self.correct_retrieval / self.query_count
+        # self.log.append({"documents": documents, "labels": labels, "correct_retrieval": correct_retrieval, **kwargs})
+        return {"documents": documents, "labels": labels, "correct_retrieval": correct_retrieval, **kwargs}, "output_1"
+
+
+class EvalReader:
+    def __init__(self):
+        self.outgoing_edges = 1
+        self.query_count = 0
+        self.top_1_em_count = 0
+        self.top_k_em_count = 0
+        self.top_1_f1_sum = 0
+        self.top_k_f1_sum = 0
+        self.top_1_em = 0.0
+        self.top_k_em = 0.0
+        self.top_1_f1 = 0.0
+        self.top_k_fi = 0.0
```

Comment:
Small typo here: should be `self.top_k_f1`

Reply:
Legend!